### ISSUE TITLE
False postives on memory leaks when built under Xcode 11.3 

### ISSUE THREAD

**Issue by richardmarion** [cpputest/cpputest#1274]
Memory leak tests that run fine on Mac OS 10.14 compiled with Xcode 10 or 11 fail when run on MacOS  10.15.2, compiled with Xcode 11.3.1 and XCode's address sanitation enabled.

**Comment by basvodde** [comment]
Uhm, I thought XCode is an IDE and not a compiler ? Could you share with me the compiler and version it runs ?

**Comment by richardmarion** [comment]
Yes, you're right.   Kinda sloppy in my writing Xcode 11.3.1  uses Apple clang version 11.0.0 (clang-1100.0.33.17)

**Comment by basvodde** [comment]
Ok, but command line on my Mac now runs 11.0.3 and shown no memory leaks when building CppUTest and running its tests (on current trunk). Could you share a bit more information? Is it still happening on trunk?

**Comment by richardmarion** [comment]
We were seeing it when we compile with Clang's Address Sanitizer on and Undefined Behavior Sanitizer on.  
The error can be mitigated by not re-using a lvalue of string type, for instance

{{std::string a = convertutf16toutf8(L"test");
CHECKEQUAL(a,"test");
a = convertutf16toutf8("L"anothertest"):
CHECKEQUAL(a,"anothertest");
can be mitigated by writing
std::string a = convertutf16toutf8(L"test");
CHECKEQUAL(a,"test");
std::string b = convertutf16toutf8("L"anothertest"):
CHECKEQUAL(b,"anothertest");}}

**Comment by basvodde** [comment]
Does it relate to the address sanitizer and undefined behavior sanitizer ?

**Comment by richardmarion** [comment]
Yes, when the address sanitizer is off, the CppUTest does not complain and all tests pass

**Comment by basvodde** [comment]
Interesting. Thanks.

**Comment by basvodde** [comment]
So... is this a duplicated of #1270 ? If so, can I close this one ?

**Comment by basvodde** [comment]
This seems very related to #1270 ? If the address sanitizer also overload operator new, then there isn't an easy solution for that. You basically got 2 pieces of code doing the same thing :)

**Comment by richardmarion** [comment]
Well, I don't recall see linker messages stating I have multiple definitions of any of the memory functions

**Comment by basvodde** [comment]
It does hijack malloc. I'll need to read the address sanitizer code to see what it does. So far, it looks very similar and I'm not surprised they don't work well together.

**Comment by basvodde** [comment]
Ok, I now managed to compile CppUTest itself with address sanitizer and I'm wondering if linking to a library that was compiler with address sanitizer will help.

If you can... would you rebuild trunk CppUTest and try that? You'll need to use the autotools build system (didn't make the CMake version yet). Then do "autoreconf -i" to generate the config file. Then "configure --enable-sanatize-address" to generate the Makefile.

Curious if it solve this.

**Comment by richardmarion** [comment]
I’ll try this out this week.

**Comment by basvodde** [comment]
Could you check if this is solved?

Also, I couldn't reproduce this as the code snippet you send couldn't be compiled. When I made it compilable, then there wasn't a problem... locally. Lemme know how I can proceed with this?

**Comment by richardmarion** [comment]
I should be able to try this out in the next week.

I’ll let you know the results then.

Rpm
